### PR TITLE
Add a normed options on fields

### LINKED ISSUE #922: Make it possible to index a field without fieldnorms
Right now indexed fields come with a fieldnorm. This fieldnorm is used for BM25 scoring.

Bm25 is not always relevant. This ticket is about adding an indexing option for which fieldnorm data is not retained.

### PR DESCRIPTION
**Pull request by lpouget**
Make fieldnorm indexation optional:

* for all types except text: added a NORMED options
* for text field:
  * if STRING, field has not fieldnorm retained
  * if TEXT, field has fieldnorm computed

### CODE REVIEW

**Comment by fulmicoton** [comment]
@lpouget  Will review next week. Sorry for having taken so much time!

File: src/schema/flags.rs [hunk 1]
```diff
@@ -30,6 +30,18 @@ pub const INDEXED: SchemaFlagList<IndexedFlag, ()> = SchemaFlagList {
     tail: (),
 };
 
+#[derive(Clone)]
+pub struct NormedFlag;
+/// Flag to mark the field as indexed.
+///
+/// The `INDEXED` flag can only be used when building `IntOptions` (`u64`, `i64` and `f64` fields)
```
Comment (fulmicoton) [inline src/schema/flags.rs]: The comment seems incorrect.

File: src/fieldnorm/mod.rs [hunk 2]
```diff
@@ -26,3 +26,97 @@ pub use self::serializer::FieldNormsSerializer;
...
+    #[ignore]
+    #[test]
+    pub fn test_fieldnorm_bug() -> crate::Result<()> {
+        let path = Path::new("test");
+        let directory: RAMDirectory = RAMDirectory::create();
+        {
+            let write: WritePtr = directory.open_write(Path::new("test"))?;
+            let serializer = FieldNormsSerializer::from_write(write)?;
+            let mut fieldnorm_writers = FieldNormsWriter::for_schema(&SCHEMA);
+            fieldnorm_writers.fill_up_to_max_doc(1u32);
```
Comment (fulmicoton) [inline src/fieldnorm/mod.rs]: this line is actually not required.

File: src/fieldnorm/mod.rs [hunk 3]
```diff
@@ -26,3 +26,97 @@ pub use self::serializer::FieldNormsSerializer;
...
+        schema_builder.build()
+    });
+
+    pub static FIELD: Lazy<Field> = Lazy::new(|| SCHEMA.get_field("field").unwrap());
+    pub static TXT_FIELD: Lazy<Field> = Lazy::new(|| SCHEMA.get_field("txt_field").unwrap());
+    pub static STR_FIELD: Lazy<Field> = Lazy::new(|| SCHEMA.get_field("str_field").unwrap());
+
+    #[ignore]
+    #[test]
+    pub fn test_fieldnorm_bug() -> crate::Result<()> {
```
Comment (fulmicoton) [inline src/fieldnorm/mod.rs]: That's an interesting use of ignore, but let's not do that.

Let's keep the test but let's give it the right values.

Comment (fulmicoton) [reply src/fieldnorm/mod.rs]: (and change its name of course)

File: src/fieldnorm/mod.rs [hunk 4]
```diff
@@ -26,3 +26,97 @@ pub use self::serializer::FieldNormsSerializer;
...
+        }
+        Ok(())
+    }
+
+    #[test]
+    pub fn test_fail_fieldnorm_cannot_registered_twice() {
+        let mut fieldnorm_writers = FieldNormsWriter::for_schema(&SCHEMA);
+        fieldnorm_writers.fill_up_to_max_doc(1u32);
+        fieldnorm_writers.record(1u32, *TXT_FIELD, 5);
+        let result = catch_unwind(AssertUnwindSafe(|| {
```
Comment (fulmicoton) [inline src/fieldnorm/mod.rs]: You could use the `#[should_panic]` attribute.

File: src/schema/flags.rs [hunk 5]
```diff
@@ -30,6 +30,18 @@ pub const INDEXED: SchemaFlagList<IndexedFlag, ()> = SchemaFlagList {
...
+/// Of course, text fields can also be indexed... But this is expressed by using either the
+/// `STRING` (untokenized) or `TEXT` (tokenized with the english tokenizer) flags.
+pub const NORMED: SchemaFlagList<NormedFlag, ()> = SchemaFlagList {
```
Comment (fulmicoton) [inline src/schema/flags.rs]: I think we do not need a flag for normed. It is a rather advanced attribute.

On the other hand, we probably want INDEXED to imply NORMED for backward compatibility

Comment (fulmicoton) [reply src/schema/flags.rs]: In other words, people who want to disable NORMED won't be able to use this syntactic sugar and will have to configure the field manually which is fine.

**Review by fulmicoton — CHANGES_REQUESTED** [review]
See comments inline.

**Comment by fmassot** [comment]
I decided to finish it as I think @lpouget has a baby to take care :)
I thought the issue was easy but... I needed to touch many parts of the code and potentially I missed some stuff.
I think the PR is almost ready, I need to add some specific tests when with fields without field norms.

@fulmicoton @PSeitz When you have the time (it's not urgent at all), you can have a look.

File: src/indexer/merger.rs [hunk 6]
```diff
@@ -41,17 +41,29 @@ use tantivy_bitpacker::minmax;
...
+                .field_type()
+                .has_fieldnorms()
+            {
+                let fieldnorms_reader = reader.get_fieldnorms_reader(field)?;
+                for doc in reader.doc_ids_alive() {
+                    let fieldnorm_id = fieldnorms_reader.fieldnorm_id(doc);
+                    count[fieldnorm_id as usize] += 1;
+                }
+            } else {
+                let segment_num_tokens = reader.inverted_index(field)?.total_num_tokens();
```
Comment (PSeitz) [inline src/indexer/merger.rs]: This estimation will work well for id fields, but we can probably do better for zipfs distribution.

File: src/indexer/merger.rs [hunk 7]
```diff
@@ -851,10 +863,11 @@ impl IndexMerger {
             segment_map[*old_doc_id as usize] = Some(new_doc_id as DocId);
         }
 
-        // The total number of tokens will only be exact when there has been no deletes.
+        // The total number of tokens will only be exact when there has been no deletes
+        // and if the field has a norm.
```
Comment (PSeitz) [inline src/indexer/merger.rs]: norm is an approximation, not exact

File: src/indexer/segment_writer.rs [hunk 8]
```diff
@@ -242,7 +242,9 @@ impl SegmentWriter {
                         )
                     };
 
-                    self.fieldnorms_writer.record(doc_id, field, num_tokens);
+                    if field_entry.has_fieldnorms() {
+                        self.fieldnorms_writer.record(doc_id, field, num_tokens);
```
Comment (PSeitz) [inline src/indexer/segment_writer.rs]: I think fieldnorms_writer should be optional and the check simply if the writer exists

File: src/schema/bytes_options.rs [hunk 9]
```diff
@@ -56,13 +71,25 @@ impl BytesOptions {
...
+impl Default for BytesOptions {
+    fn default() -> BytesOptions {
+        BytesOptions {
+            indexed: false,
+            fieldnorms: false,
+            fast: false,
+            stored: false,
+        }
+    }
+}
```
Comment (PSeitz) [inline src/schema/bytes_options.rs]: why not use the derive? seem to be the same values

File: src/schema/text_options.rs [hunk 10]
```diff
@@ -69,6 +71,17 @@ impl TextFieldIndexing {
         &self.tokenizer
     }
 
+    /// Sets fieldnorms TODO
```
Comment (PSeitz) [inline src/schema/text_options.rs]:
```suggestion
    /// Sets fieldnorms
```

File: src/schema/text_options.rs [hunk 11]
```diff
@@ -69,6 +71,17 @@ impl TextFieldIndexing {
...
+    pub fn set_fieldnorms(mut self, fieldnorms: bool) -> TextFieldIndexing {
+        self.fieldnorms = fieldnorms;
+        self
+    }
+
+    /// Sets fieldnorms TODO
```
Comment (PSeitz) [inline src/schema/text_options.rs]:
```suggestion
    /// Returns true iff fieldnorms are stored.
```

File: src/schema/bytes_options.rs (hunk 9 above)
Comment (fmassot) [reply src/schema/bytes_options.rs]: yes!

File: src/indexer/merger.rs (hunk 7 above)
Comment (fulmicoton) [reply src/indexer/merger.rs:867]: Yes but this is not fieldnorm here.

**Comment by fulmicoton** [comment]
@lpouget I will take over that PR and merge it. I hope this is ok with you.

File: src/indexer/segment_writer.rs (hunk 8 above)
Comment (fulmicoton) [reply src/indexer/segment_writer.rs]: Agreed

**Comment by fulmicoton** [comment]
This PR was merged as branch issue/922b

**Pull request closed without merging**